### PR TITLE
Fix moving hero animation priority

### LINKED ISSUE #1293: Display priority of the hero's sprite during the movement animation in relation to another hero
While moving, the blue hero's sprite is displayed under the green hero's sprite.

Display priority of the hero's sprite during the movement animation in relation to another hero 01

Here is the test scenario file:
[Abandoned mine no victory window.zip](https://github.com/ihhub/fheroes2/files/5098267/Abandoned.mine.no.victory.window.zip)

Set the Hero Speed to 1.
Move the first hero from the list to go to the abandoned mine.

### PR DESCRIPTION
**Pull request by vasilenkoalexey**
Fix #1293

### CODE REVIEW

File: src/fheroes2/gui/interface_gamearea.cpp [hunk 1]
```diff
@@ -295,6 +295,32 @@ void Interface::GameArea::Redraw( fheroes2::Image & dst, int flag, bool isPuzzle
         }
     }
 
+    if ( drawList.size() > 0 ) {
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/gui/interface_gamearea.cpp:298]: We should compare this number with 1, not 0. Only 2 or more heroes in the container are needed for sorting.

File: src/fheroes2/gui/interface_gamearea.cpp [hunk 2]
```diff
@@ -295,6 +295,32 @@ void Interface::GameArea::Redraw( fheroes2::Image & dst, int flag, bool isPuzzle
...
+        for ( auto it = drawList.begin(); it != drawList.end() - 1; ) {
```
Comment (ihhub) [inline src/fheroes2/gui/interface_gamearea.cpp:299]: Could we please just use a normal counter-based loop here? Reading iterator-based code is really hard.

File: src/fheroes2/gui/interface_gamearea.cpp [hunk 3]
```diff
@@ -295,6 +295,32 @@ void Interface::GameArea::Redraw( fheroes2::Image & dst, int flag, bool isPuzzle
...
+            Heroes * hero1 = ( *it )->GetHeroes();
```
Comment (ihhub) [inline src/fheroes2/gui/interface_gamearea.cpp:300]: Please declare here a new variable called `const Maps::Tiles * tile = drawList[i];` as it below.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @vasilenkoalexey , I left few comments here. Would you mind to take a look?

File: src/fheroes2/gui/interface_gamearea.cpp [hunk 4]
```diff
@@ -295,6 +295,32 @@ void Interface::GameArea::Redraw( fheroes2::Image & dst, int flag, bool isPuzzle
...
+            if ( ( hero2 == nullptr && MP2::OBJ_BOAT != object ) || !Maps::isValidDirection( ( *it )->GetIndex(), Direction::RIGHT ) ) {
+                ++it;
+                continue;
+            }
+            const Maps::Tiles & tileRight = world.GetTiles( Maps::GetDirectionIndex( ( *it )->GetIndex(), Direction::RIGHT ) );
+            if ( tileRight.GetIndex() != ( *( it + 1 ) )->GetIndex() ) {
+                ++it;
+                continue;
+            }
+            std::swap( *it, *( it + 1 ) );
```
Comment (ihhub) [inline src/fheroes2/gui/interface_gamearea.cpp:320]: Do we need to break the loop here? We know that only 1 hero is moving at once.

**Comment by Branikolog** [comment]
It seems that hero overlaps another one standing nearby While moving northwest..
image

**Comment by ihhub** [comment]
Hi @vasilenkoalexey , I think we need to priotirized hero animation only for a situation of a hero moving to bottom-left direction. To do this we might need to add a special bollean flag in Hero class to indicate that the hero is moving.

**Comment by vasilenkoalexey** [comment]
Also should fix #2500

File: src/fheroes2/gui/interface_gamearea.cpp [hunk 5]
```diff
@@ -295,6 +302,24 @@ void Interface::GameArea::Redraw( fheroes2::Image & dst, int flag, bool isPuzzle
         }
     }
 
+    if ( drawList.size() > 1 && tileMovingHero != nullptr ) {
+        const Heroes * hero = tileMovingHero->GetHeroes();
+        if ( hero->GetDirection() == Direction::BOTTOM_RIGHT ) {
+            const Maps::Tiles & tileNeighborHero = world.GetTiles( Maps::GetDirectionIndex( tileMovingHero->GetIndex(), Direction::RIGHT ) );
+            if ( tileNeighborHero.GetHeroes() != nullptr || tileNeighborHero.GetObject() == MP2::OBJ_BOAT ) {
+                const auto it = std::find( drawList.begin(), drawList.end(), tileMovingHero );
+                std::iter_swap( it, it + 1 );
```
Comment (ihhub) [inline src/fheroes2/gui/interface_gamearea.cpp:311]: Should we check the next hero in the list and his tile ID?
```
const auto currentHeroTile = std::find( drawList.begin(), drawList.end(), tileMovingHero );
assert( currentHeroTile != drawList.end() );
const auto nextHeroTile = currentHeroTile + 1;
if ( nextHeroTile != drawList.end() && ( nextHeroTile->GetIndex() == currentHeroTile->GetIndex() + 1 )
    && ( nextHeroTile.GetObject() == MP2::OBJ_HEROES || nextHeroTile.GetObject() == MP2::OBJ_BOAT ) ) {
    std::iter_swap( currentHeroTile, nextHeroTile );
}
```

The same logic applies for the case below.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @vasilenkoalexey , I left one comment regarding logic. Could you please take a look?

File: src/fheroes2/gui/interface_gamearea.cpp (hunk 5 above)
Comment (vasilenkoalexey) [reply src/fheroes2/gui/interface_gamearea.cpp:311]: We process tiles from top to bottom and left to RIGHT, its mean we put heroes to list in same order. Here, we get neighbor hero exactly at RIGHT tile(in relation to the current hero), and its mean that current hero tile index is one less than neighbor tile index, and neighbor hero should be always next in heroes list (again in relation to the current hero). So I think additional checks are unnecessary here.

Comment (ihhub) [reply src/fheroes2/gui/interface_gamearea.cpp:311]: Additional checks will help to verify the described logic by you. If someone else, even me, will change a piece of the code like sort heroes and etc. then at least we can catch the incorrect behavior. Nothing more than this. We can repalce `nextHeroTile->GetIndex() == currentHeroTile->GetIndex() + 1` by `nextHeroTile->GetIndex() == Maps::GetDirectionIndex( currentHeroTile->GetIndex(), Direction::RIGHT )` to keep even better readability without hardcoding.

Comment (Branikolog) [reply src/fheroes2/gui/interface_gamearea.cpp]: While moving northwest hero is still overlapping a standing hero.
Bug

**Comment by ihhub** [comment]
I am closing this pull request as the upcoming Adventure Map rendering would cover this issue.

**Pull request closed without merging**